File: Engine/Leviathan/Window.cs
using Leviathan.Configuration;
using Leviathan.Configuration.Converters.Math;
using Leviathan.Events;
using Leviathan.GameObjects;
using Leviathan.Mathematics;

using Raylib_cs;

using Color = Leviathan.Mathematics.Color;

namespace Leviathan
{
	public sealed class Window : IEventHandler
	{
		public ref Vector2Int ScreenSize => ref screenSize;
		public ref Color ClearColor => ref clearColor;
		public bool IsFullscreen { get; private set; }
		public string? Title { get; private set; }
		public Rectangle ScreenBounds => new(0, 0, screenSize.x, screenSize.y);

		private Vector2Int screenSize;
		private Color clearColor;
		private long frameRate;
		private bool lockFrameRate;
		private long fullscreenKey;

		private readonly Config<ApplicationConfigData>? programConfig;

		internal Window(Config<ApplicationConfigData>? _programConfig)
		{
			programConfig = _programConfig ?? throw new NullReferenceException($"Program config is null!");
			
			screenSize = programConfig.GetValue("window.screenSize", Int2Converter.Instance);
			Title = programConfig.GetValue<string>("application.name");
			clearColor = programConfig.GetValue("application.clearColor", ColorConverter.Instance);
			IsFullscreen = programConfig.GetValue<bool>("window.useFullscreen");
			fullscreenKey = programConfig.GetValue<long>("window.fullscreenKey");
			frameRate = programConfig.GetValue<long>("application.maxFrameRate");
			lockFrameRate = programConfig.GetValue<bool>("application.lockFrameRate");
			
			if(lockFrameRate)
				Raylib.SetTargetFPS((int)frameRate);
		}

		public void ClearScreen() => Raylib.ClearBackground(clearColor);

		internal void Open()
		{
			EventBus.RegisterObject(this);
			
			Application.raylibLogger.LogWarn($"Opening window with size {screenSize} and name {Title}");
			Raylib.InitWindow(screenSize.x, screenSize.y, Title!);
			
			if(IsFullscreen)
			{
				screenSize = new Vector2Int
				{
					x = Raylib.GetMonitorWidth(0),
					y = Raylib.GetMonitorHeight(0),
				};
				
				Raylib.ToggleFullscreen();
				Raylib.SetWindowSize(screenSize.x, screenSize.y);
			}
		}

		internal void Close()
		{
			Application.raylibLogger.LogWarn($"Closing window with size {screenSize} and name {Title}");
			Raylib.CloseWindow();
			
			EventBus.RemoveObject(this);
		}

		internal void Render()
		{
			Raylib.BeginDrawing();
			ClearScreen();
			
			GameObjectManager.Render();
			
			Raylib.EndDrawing();
		}

		internal void Tick()
		{
			if(Raylib.IsKeyPressed((KeyboardKey) fullscreenKey))
				SwapFullscreenMode();
		}

		public void SwapFullscreenMode()
		{
			if(programConfig == null)
				throw new NullReferenceException("Program config is null!");
			
			IsFullscreen = !IsFullscreen;
			
			if(!IsFullscreen)
				Raylib.ToggleFullscreen();

			Vector2Int size = programConfig.GetValue("window.screenSize", Int2Converter.Instance);
			screenSize = new Vector2Int
			{
				x = IsFullscreen ? Raylib.GetMonitorWidth(0) : size.x,
				y = IsFullscreen ? Raylib.GetMonitorHeight(0) : size.y,
			};
			Raylib.SetWindowSize(screenSize.x, screenSize.y);
			
			if(IsFullscreen)
				Raylib.ToggleFullscreen();
		}

		[SubscribeEvent]
		// ReSharper disable once UnusedMember.Local
		// ReSharper disable once UnusedParameter.Local
		private void Reconfigure(ConfigReloadEvent _event)
		{
			if(programConfig == null)
				throw new NullReferenceException("Program config is null!");

			bool wasFullscreen = IsFullscreen;

			screenSize = programConfig.GetValue("window.screenSize", Int2Converter.Instance);
			Title = programConfig.GetValue<string>("application.name");
			clearColor = programConfig.GetValue("application.clearColor", ColorConverter.Instance);
			IsFullscreen = programConfig.GetValue<bool>("window.useFullscreen");
			fullscreenKey = programConfig.GetValue<long>("window.fullscreenKey");
			frameRate = programConfig.GetValue<long>("application.maxFrameRate");
			lockFrameRate = programConfig.GetValue<bool>("application.lockFrameRate");
			
			Raylib.SetWindowTitle(Title!);
			Raylib.SetWindowSize(screenSize.x, screenSize.y);
			if(lockFrameRate)
				Raylib.SetTargetFPS((int)frameRate);

			if(!wasFullscreen && IsFullscreen)
			{
				Raylib.ToggleFullscreen();
				screenSize = new Vector2Int
				{
					x = Raylib.GetMonitorWidth(0),
					y = Raylib.GetMonitorHeight(0),
				};
				Raylib.SetWindowSize(screenSize.x, screenSize.y);
			}
			else if(wasFullscreen && !IsFullscreen)
			{
				Raylib.ToggleFullscreen();
				Raylib.SetWindowSize(screenSize.x, screenSize.y);
			}
		}
	}
}

File: Framework/Leviathan/Core/Window.cpp
#include <Leviathan/Window.h>

#include <Leviathan/Application.h>

#include <Leviathan/Utils/Config.h>
#include <Leviathan/Utils/ConfigValue.h>

#include <iostream>

#include <raylib/raylib.h>

namespace Leviathan
{
    Window::Window()
        : m_width(0), m_height(0), m_fullscreenKey(0), m_config(nullptr)
    {
    }

    void Window::Open()
    {
        m_config = Application::GetConfig(ConfigType::Window);

        m_width = m_config->GetValue(WINDOW_CATEGORY, "width")->Get<int>();
        m_height = m_config->GetValue(WINDOW_CATEGORY, "height")->Get<int>();
        m_title = m_config->GetValue(WINDOW_CATEGORY, "title")->Get<string>();
        m_clearColor = m_config->GetValue(WINDOW_CATEGORY, "clearColor")->Get<Color32>();
        m_fullscreenKey = m_config->GetValue(WINDOW_CATEGORY, "toggleFullscreenKey")->Get<int>();
        m_config->ListenForReload(&Window::OnConfigReloaded, this);

        if (m_config->GetValue(WINDOW_CATEGORY, "useFullscreen")->Get<bool>())
        {
            SetConfigFlags(FLAG_FULLSCREEN_MODE);
            m_width = m_height = 0;
        }
        else
        {
            SetConfigFlags(FLAG_WINDOW_RESIZABLE);
        }

        InitWindow(m_width, m_height, m_title.c_str());

    }

    void Window::Close()
    {
        CloseWindow();
    }

    void Window::BeginFrame()
    {
        TryToggleFullscreen();

        BeginDrawing();
        ClearBackground(m_clearColor);

        m_width = GetScreenWidth();
        m_height = GetScreenHeight();
    }

    void Window::EndFrame()
    {
        EndDrawing();
    }

    void Window::TryToggleFullscreen()
    {
        if (IsKeyPressed(m_fullscreenKey))
        {
            if (!IsWindowFullscreen())
            {
                SetWindowSize(GetMonitorWidth(0), GetMonitorHeight(0));
            }

            ToggleFullscreen();

            if (!IsWindowFullscreen())
            {
                int width = m_config->GetValue(WINDOW_CATEGORY, "width")->Get<int>();
                int height = m_config->GetValue(WINDOW_CATEGORY, "height")->Get<int>();

                SetWindowSize(width, height);
            }
        }
    }

    void Window::OnConfigReloaded(Config* _config)
    {
        m_width = m_config->GetValue(WINDOW_CATEGORY, "width")->Get<int>();
        m_height = m_config->GetValue(WINDOW_CATEGORY, "height")->Get<int>();
        m_title = m_config->GetValue(WINDOW_CATEGORY, "title")->Get<string>();
        m_clearColor = m_config->GetValue(WINDOW_CATEGORY, "clearColor")->Get<Color32>();
        m_fullscreenKey = m_config->GetValue(WINDOW_CATEGORY, "toggleFullscreenKey")->Get<int>();

        SetWindowTitle(m_title.c_str());
        if (!IsWindowFullscreen())
        {
            SetWindowSize(m_width, m_height);
        }
    }
}